in-app-purchase/src/main/java/com/rakuten/tech/mobile/miniapp/iap/InAppPurchaseExecutor.kt

```diff
@@ -0,0 +1,127 @@
+import android.util.Log
+import com.android.billingclient.api.BillingClient
+import com.android.billingclient.api.BillingClientStateListener
+import com.android.billingclient.api.BillingFlowParams
+import com.android.billingclient.api.BillingResult
+import com.android.billingclient.api.ConsumeParams
+import com.android.billingclient.api.Purchase
+import com.android.billingclient.api.PurchasesUpdatedListener
+import com.android.billingclient.api.SkuDetails
+import com.android.billingclient.api.SkuDetailsParams
+
+/**
+ * The In-App purchase executor which acts as a default implementation of InAppPurchaseProvider.
+ * @param context should use the same activity context for #MiniAppDisplay.getMiniAppView.
+ */
+class InAppPurchaseExecutor(private val context: Activity) : InAppPurchaseProvider {
+    private var billingClient: BillingClient? = null
+    private var skuDetails: SkuDetails? = null
+
+    private fun startPurchasingProduct(itemID: String) {
+        setUpBillingClient(itemID)
+        retrieveSkuDetails()
+    }
+
+    private fun setUpBillingClient(productID: String) {
+        billingClient = BillingClient.newBuilder(context)
+            .setListener(purchaseUpdateListener)
+            .enablePendingPurchases()
+            .build()
+
+        startConnection(productID)
+    }
+
+    private fun startConnection(productID: String) {
+        billingClient?.startConnection(object : BillingClientStateListener {
+            override fun onBillingSetupFinished(billingResult: BillingResult) {
+                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
+                    queryAvailableProducts(productID)
+                }
+            }
+
+            override fun onBillingServiceDisconnected() {
+                Log.d(TAG, "Billing service has been disconnected.")
+            }
+        })
+    }
+
+    private fun queryAvailableProducts(productID: String) {
+        val skuList = ArrayList<String>()
+        skuList.add(productID)
+        val params = SkuDetailsParams.newBuilder()
+
+        // proceed with In-App type
+        params.setSkusList(skuList).setType(BillingClient.SkuType.INAPP)
+
+        billingClient?.querySkuDetailsAsync(params.build()) { billingResult, skuDetailsList ->
+            // Process the result.
+            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK && !skuDetailsList.isNullOrEmpty()) {
+                for (skuDetails in skuDetailsList) {
+                    this.skuDetails = skuDetails
+                }
+            } else {
+                Log.e(TAG, "There is an error happened while purchasing item")
+            }
+        }
+    }
+
+    private fun retrieveSkuDetails() {
+        skuDetails?.let {
+            val billingFlowParams = BillingFlowParams.newBuilder()
+                .setSkuDetails(it)
+                .build()
+            billingClient?.launchBillingFlow(context, billingFlowParams)?.responseCode
```

Comment:
Looks like there needs to be some handling for the `responseCode` here. According to the documentation, normally the response is delivered to the `PurchaseUpdatedListener`, but there are some cases where that listener would not be called and you would instead need to check this responseCode (`ITEM_ALREADY_OWNED`, `ITEM_UNAVAILABLE`, `USER_CANCELED`): https://developer.android.com/reference/com/android/billingclient/api/BillingClient#launchBillingFlow(android.app.Activity,%20com.android.billingclient.api.BillingFlowParams)

So we should at least check here if it's `OK` and then send an `onError` response to the Mini App if not.

Reply:
Yeah, I have added those responseCode e.g. `ITEM_ALREADY_OWNED`, `ITEM_UNAVAILABLE`, `USER_CANCELED` to be invoked by `onError`.